lib/embulk/input/twitter_ads/api.rb

```diff
@@ -0,0 +1,80 @@
+require "active_support"
+require "active_support/core_ext/numeric"
+
+module Embulk
+  module Input
+    module TwitterAds
+      module API
+        NUMBER_OF_RETRIES = 5
+        MAX_SLEEP_SEC_NUMBER = 1200
+
+        # Error codes and responses
+        # @see https://developer.twitter.com/en/docs/twitter-ads-api/response-codes
+        # Client Errors (4XX)
+        class ClientError < StandardError; end
+        class BadRequest < ClientError; end
+        class NotAuthorized < ClientError; end
+        class Forbidden < ClientError; end
+        class NotFound < ClientError; end
+        class RateLimit < ClientError; end
+        # Server Errors (5XX)
+        class ServerError < StandardError; end
+        class ServiceUnavailable < ServerError; end
+
+        ERRORS = {
+          "400" => BadRequest,
+          "401" => NotAuthorized,
+          "403" => Forbidden,
+          "404" => NotFound,
+          "429" => RateLimit,
+          "500" => ServerError,
+          "503" => ServiceUnavailable
+        }.freeze
+
+        def with_retry(logger:)
+          retries = 0
+          begin
+            response = yield
+            if ERRORS["#{response.code}"].present?
+              logger.error "#{response.body}"
+              raise ERRORS["#{response.code}"]
+            end
+            return response
+          rescue RateLimit, ServerError => e
+            if retries < NUMBER_OF_RETRIES
+              retries += 1
+              sleep_sec = get_sleep_sec(response: response, retries: retries)
+              if sleep_sec > MAX_SLEEP_SEC_NUMBER
+                raise e
+              end
+              logger.info "waiting for retry #{sleep_sec} seconds"
+              sleep sleep_sec
+              logger.warn("retry #{retries}, #{e.message}")
+              retry
+            else
+              logger.error("exceeds the upper limit retry, #{e.message}")
+              raise e
+            end
+          end
+        end
+
+        private
+        def get_sleep_sec(response:, retries:)
```

Comment:
![memo-badge](https://img.shields.io/badge/review-memo-lightgrey)

duplicated with the following code:
https://github.com/trocco-io/embulk-input-twitter_ads_analytics/blob/44dd25302f46746f4fa670322b67a0b0e4bd52d5/lib/embulk/input/twitter_ads_analytics.rb#L408-L422